ports/x11/verbiste/verbiste-0.1.41/doc/french-deconjugator.1
.\" $Id: french-deconjugator.1.in,v 1.12 2013/07/18 03:13:05 sarrazip Exp $
.\" french-deconjugator - Analyzer of conjugated French verbs
.\" verbiste - French conjugation system
.\" Copyright (C) 2003-2009 Pierre Sarrazin <http://sarrazip.com/>
.\"
.\" This program is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License
.\" as published by the Free Software Foundation; either version 2
.\" of the License, or (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
.\" 02110-1301, USA.
.\"
.\"
.TH french-deconjugator "1" "April 6th, 2014" "" ""
.SH NAME
french-deconjugator - analyze conjugated French verbs
.SH SYNOPSIS
.B echo aim\('e | french-deconjugator > result.txt
.SH DESCRIPTION
\fBfrench-deconjugator\fR
reads conjugated French verbs from
the command line or from standard input
and writes (to standard output)
the verb's \fBinfinitive\fR form,
the \fBmode\fR (infinitive, indicative, conditional, subjunctive,
imperative or participle),
the \fBtense\fR (present, past, imperfect, future),
the \fBperson\fR (1, 2 or 3,
while 0 is used for the present participle tense,
and 4 and 5 are used in the past participle tense),
and
the \fBnumber\fR (singular or plural).
These fields are separated by a comma and a space.
.PP
The standard input is not read if verbs are passed as command-line
arguments.
.PP
By convention, persons 4 and 5 are used in the
\fBpast participle\fR
tense to indicate the gender:
4 means masculine (e.g., "aim\('e" or "aim\('es")
and 5 means feminine (e.g., "aim\('ee" or "aim\('ees").
.PP
A single conjugated form can correspond to more than one mode,
tense and person.
In this case, each alternative is written on its own line.
.PP
In all cases, the end of the answer is marked by an empty line.
If the word is unknown, only this empty line is written.
The names for the mode, tense and number are always in English.
(This is meant to facilitate automatic parsing of the output.
For a French user interface, see the GNOME application and applet.)
.PP
The command flushes its output buffer after finishing each answer.
This allows the command to be easily called from another program
through two pipes.
.PP
The command starts by loading its database from XML files
(stored typically in /usr/share/verbiste).
This takes some time, so it is a good idea to have the command
answer many requests instead of running it for each request.
.PP
The verbiste library's source archive contains Perl and Java
example programs that illustrate this technique.
.PP
This commands expects to read Latin-1 characters and writes Latin-1 characters.
There must not be any leading or trailing white spaces on the lines read
by the command.
.SH OPTIONS
.TP
\fB\-\-help\fR
display a help page and exit
.TP
\fB\-\-version\fR
display version information and exit
.TP
\fB\-\-lang=L\fR
select the language to use (\fBfr\fR for French or \fBit\fR for Italian);
French is the default language
.TP
\fB\-\-all-infinitives\fR
print the infinitive form of all the verbs in the knowledge base,
one per line, unsorted; other command-line arguments are ignored
.SH EXAMPLES
$ french-deconjugator aim\('e
.br
aimer, participle, past, 0, singular
.br

.br
$ echo \-ne 'a\enplu\en\('et\('e\en' | french-deconjugator
.br
avoir, indicative, present, 3, singular

.br
plaire, participle, past, 0, singular
.br
pleuvoir, participle, past, 0, singular
.br

.br
\(^etre, participle, past, 0, singular
.br

.br
.SH LICENSE
This program is free software; you may redistribute it under the terms of
the GNU General Public License.  This program has absolutely no warranty.
.SH AUTHOR
See the verbiste(3) manual page.
.SH BUGS
See the verbiste(3) manual page.
.SH SEE ALSO
verbiste(3), french-conjugator(1).


ports/x11/verbiste/verbiste-0.1.41/src/commands/french-deconjugator.cpp
/*  $Id: french-deconjugator.cpp,v 1.25 2012/11/18 19:56:05 sarrazip Exp $
    french-deconjugator.cpp - Analyzer of conjugated French verbs

    verbiste - French conjugation system
    Copyright (C) 2003-2010 Pierre Sarrazin <http://sarrazip.com/>

    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public License
    as published by the Free Software Foundation; either version 2
    of the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
    02110-1301, USA.
*/


#include "Command.h"

#ifdef ENABLE_NLS
#ifdef HAVE_GETOPT_LONG
#include <unistd.h>
#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif
#include <getopt.h>
#endif
#endif  /* ENABLE_NLS */

#include <langinfo.h>

#include <iostream>
#include <vector>
#include <string.h>

using namespace std;
using namespace verbiste;


static const char *commandName = "french-deconjugator";


#if defined(ENABLE_NLS) && defined(HAVE_GETOPT_LONG)
static struct option knownOptions[] =
{
    { "help",            no_argument,       NULL, 'h' },
    { "version",         no_argument,       NULL, 'v' },
    { "lang",            required_argument, NULL, 'l' },
    { "all-infinitives", no_argument,       NULL, 'i' },

    { NULL, 0, NULL, 0 }  // marks the end
};
#endif


class DeconjugatorCommand : public Command
{
public:

    DeconjugatorCommand(const string &conjugationFilename,
                        const string &verbsFilename,
                        FrenchVerbDictionary::Language lang) throw(logic_error)
      : Command(conjugationFilename, verbsFilename, lang)
    {
    }

    virtual ~DeconjugatorCommand()
    {
    }

protected:

    virtual void processInputWord(const std::string &inputWord);

};


/*virtual*/
void
DeconjugatorCommand::processInputWord(const string &inputWord)
{
    // Analyze the word and get the results in a vector:
    vector<InflectionDesc> v;
    fvd->deconjugate(inputWord, v);


    for (vector<InflectionDesc>::const_iterator it = v.begin();
                                            it != v.end(); it++)
    {
        const InflectionDesc &d = *it;
        cout
            << d.infinitive
            << ", " << FrenchVerbDictionary::getModeName(d.mtpn.mode)
            << ", " << FrenchVerbDictionary::getTenseName(d.mtpn.tense)
            << ", " << int(d.mtpn.person)
            << ", " << (d.mtpn.plural ? "plural" : "singular")
            << "\n";
    }
    cout << endl;
}


static
void
displayVersionNo()
{
    cout << commandName << ' ' << VERSION << '\n';
}


static
void
displayHelp()
{
    cout << '\n';

    displayVersionNo();

    cout << "Part of " << PACKAGE << " " << VERSION << "\n";

    cout <<
"\n"
"Copyright (C) " COPYRIGHT_YEARS " Pierre Sarrazin <http://sarrazip.com/>\n"
"This program is free software; you may redistribute it under the terms of\n"
"the GNU General Public License.  This program has absolutely no warranty.\n"
    ;

    cout <<
"\n"
"Options:\n"
"--help             Display this help page and exit\n"
"--version          Display this program's version number and exit\n"
"--lang=L           Select language L (fr for French, it for Italian)\n"
"                   Default is French.\n"
"--all-infinitives  Print the names of all known verbs, one per line (unsorted)\n"
"\n"
"See the " << commandName << "(1) manual page for details.\n"
"\n"
    ;
}


int
main(int argc, char *argv[])
{
    setlocale(LC_CTYPE, "");
    bool listAllInfinitives = false;
    string langCode = "fr";

    #if defined(ENABLE_NLS) && defined(HAVE_GETOPT_LONG)

    /*  Interpret the command-line options:
    */
    int c;
    do
    {
        c = getopt_long(argc, argv, "hv", knownOptions, NULL);

        switch (c)
        {
            case EOF:
                break;  // nothing to do

            case 'l':
                langCode = optarg;
                if (FrenchVerbDictionary::parseLanguageCode(langCode) == FrenchVerbDictionary::NO_LANGUAGE)
                {
                    cerr << commandName << ": invalid language code " << optarg << "\n";
                    return EXIT_FAILURE;
                }
                break;

            case 'v':
                displayVersionNo();
                return EXIT_SUCCESS;

            case 'h':
                displayHelp();
                return EXIT_SUCCESS;

            case 'i':
                listAllInfinitives = true;
                break;

            default:
                displayHelp();
                return EXIT_FAILURE;
        }
    } while (c != EOF && c != '?');

    #else

    int optind = 1;

    #endif

    try
    {
        FrenchVerbDictionary::Language lang = FrenchVerbDictionary::parseLanguageCode(langCode);
        if (lang == FrenchVerbDictionary::NO_LANGUAGE)
            lang = FrenchVerbDictionary::FRENCH;

        string conjFN, verbsFN;
        FrenchVerbDictionary::getXMLFilenames(conjFN, verbsFN, lang);

        DeconjugatorCommand cmd(conjFN, verbsFN, lang);

        if (listAllInfinitives)
            return cmd.listAllInfinitives(cout);

        return cmd.run(argc - optind, argv + optind);
    }
    catch (const exception &e)
    {
        cerr << commandName << ": exception: " << e.what() << endl;
        return EXIT_FAILURE;
    }
}
